Userland/Libraries/LibCore/FileWatcher.cpp

```diff
@@ -24,36 +25,141 @@
+        return {};
+    }
+    String const& path = it->value;
+
+    switch (event->type) {
+    case InodeWatcherEvent::Type::ChildCreated:
+        result.type = FileWatcherEvent::Type::ChildCreated;
+        break;
+    case InodeWatcherEvent::Type::ChildDeleted:
+        result.type = FileWatcherEvent::Type::ChildDeleted;
+        break;
+    case InodeWatcherEvent::Type::Deleted:
+        result.type = FileWatcherEvent::Type::Deleted;
+        break;
+    case InodeWatcherEvent::Type::ContentModified:
+        result.type = FileWatcherEvent::Type::ContentModified;
+        break;
+    case InodeWatcherEvent::Type::MetadataModified:
+        result.type = FileWatcherEvent::Type::MetadataModified;
+        break;
+    default:
+        warnln("Unknown event type {} returned by the watch_file descriptor for {}", static_cast<unsigned>(event->type), path);
+        return {};
+    }
 
-        struct stat st = {};
-        if (lstat(child_full_path.characters(), &st) < 0) {
+    // We trust that the kernel only sends the name when appropriate.
+    if (event->name_length > 0) {
+        String child_name { event->name, event->name_length - 1 };
+        auto lexical_path = LexicalPath::join(path, child_name);
+        if (!lexical_path.is_valid()) {
+            dbgln_if(FILE_WATCHER_DEBUG, "get_event_from_fd: Reading from wd {}: Invalid child name '{}'", fd, child_name);
             return {};
         }
 
-        if (st.st_ino == child_inode_index) {
-            return child_full_path;
-        }
+        result.event_path = lexical_path.string();
+    } else {
+        result.event_path = path;
     }
-    return {};
+
+    dbgln_if(FILE_WATCHER_DEBUG, "get_event_from_fd: got event from wd {} on '{}' type {}", fd, result.event_path, result.type);
+    return result;
 }
 
-BlockingFileWatcher::BlockingFileWatcher(const String& path)
-    : m_path(path)
+Result<bool, String> FileWatcherBase::add_watch(String path, InodeWatcherEvent::Type event_mask)
+{
+    LexicalPath lexical_path;
+    if (path.length() > 0 && path[0] == '/') {
+        lexical_path = LexicalPath { path };
+    } else {
+        char* buf = getcwd(nullptr, 0);
+        lexical_path = LexicalPath::join(String(buf), path);
+        free(buf);
+    }
+
+    if (!lexical_path.is_valid()) {
+        dbgln_if(FILE_WATCHER_DEBUG, "add_watch: path '{}' invalid", path);
+        return false;
+    }
+
+    auto const& canonical_path = lexical_path.string();
+    if (m_path_to_wd.find(canonical_path) != m_path_to_wd.end()) {
+        dbgln_if(FILE_WATCHER_DEBUG, "add_watch: path '{}' is already being watched", canonical_path);
+        return false;
+    }
+
+    int wd = iwatcher_add_watch(m_watcher_fd, canonical_path.characters(), canonical_path.length(), static_cast<unsigned>(event_mask));
+    if (wd < 0)
+        return String::formatted("{}: Could not watch file '{}' : {}", class_name(), canonical_path, strerror(errno));
```

Comment:
Should this error message be prefixed with `add_watch:` like all other errors in this function?